Streamline key management for build caches

opadron opened the pull request:
Fixes the logic Spack uses when getting keys from a mirror, which was broken after merging #15002 and #15361.

- ~Update `spack.util.web.list_url()` to iterate recursively over all files~
   - ~Changed to be more consistent across `s3://` and `file://` URLs.~

 - ~Changed `bindist.get_keys()` so that it no longer depends on an index.~
   - ~Instead, it will find *all* files under the prefix URL that end
     with .key or .pub~

 - Simplified `bindist.get_keys()` so that it uses the web/url abstraction
   instead of separate code paths for `s3://` and `file://` resources.

TODO:

- [x] Fix tests
- [x] Confirm getting keys from mirror is working
- [x] Add new tests covering changes

EDIT: Crossed out parts of this PR that has since changed.  See discussion for details.

Commit:
Simplify bindist.get_keys()

 - Update spack.util.web.list_url() to iterate recursively over all files
   - Changed to be more consistent across s3:// and file:// URLs.

 - Changed bindist.get_keys() so that it no longer depends on an index.
   - Instead, it will find *all* files under the prefix URL that end
     with .key or .pub

 - Simplified bindist.get_keys() so that it uses the web/url abstraction
   instead of separate code paths for s3:// and file:// resources.

Commit:
fix tests

Commit:
search for keys starting at mirror root

opadron commented:
Tagging with "don't-merge-yet": There is still more work to be done here with respect to basic key management. (per conversations with @tgamblin)

scottwittenburg commented:
@opadron Quick question: It looks like this is only dealing with the read side of things, do we already have nice commands for adding, say, a public key to a buildcache?

Attn: @eugeneswalker

gartung commented:
I never added that ability add a public key to the buildcache tarfile.

I also have an issue opened to add signing of buildcaches after they are created.
https://github.com/spack/spack/issues/13732

gartung commented:
Or better yet, move the downloading and trusting of keys to the spack gpg command.

scottwittenburg commented:
> I never added that ability add a public key to the buildcache tarfile.

Sorry I wasn't clear there.  I meant to ask about just uploading a public key to the buildcache directory, for example, into the directory where this PR will look for them.  I think in a recent discussion were thinking of expecting keys to live under the buildcache directory in a folder named `_keys`.  That folder would also contain a simple json index of the available keys.  And the command we might introduce for adding such a key would also handle updating that index.

eugeneswalker commented:
I'm not sure I understand exactly the plan here, but if it is to list all objects in a directory that could potentially contain many thousands of files in an attempt to find those that have .key or .pub extensions, I am against that. Are we talking about a specific directory like `_keys` where only keys would be stored?

scottwittenburg commented:
> the command we might introduce for adding such a key would also handle updating that index.

It might also be nice, when we're adding a buildcache entry to a mirror via `spack buildcache create`, to just make sure the public part of the key is uploaded and indexed as well.

scottwittenburg commented:
> Are we talking about a specific directory like `_keys` where only keys would be stored?

Yes, IIRC that's currently the plan, although I don't see that notion reflected here yet.

opadron commented:
That's right.  The plan for this PR is to pivot on the original idea and move to a `_keys` directory under the build cache that will have all the keys and a json index.  This PR will also include plumbing for uploading the keys to the build cache (and subsequently updating the index).

opadron commented:
For `spack.util.web.list_url()`, the changes are moot for this PR since we're moving away from programmatic listings and relying on an index.  I will probably end up reverting that part of the change set.

Commit:
Rework spack.util.web.list_url()

list_url() now accepts an optional recursive argument (default: False)
for controlling whether to only return files within the prefix url or to
return all files whose path starts with the prefix url.  Allows for the
most effecient implementation for the given prefix url scheme.  For
example, only recursive queries are supported for S3 prefixes, so the
returned list is trimmed down if recursive == False, but the native
search is returned as-is when recursive == True.  Suitable
implementations for each case are also used for file system URLs.

This (partially) reverts commit 528365da5b3333f79ee3694266f930f1b7f72a03.

Commit:
Switch to using an explicit index for public keys

Switches to maintaining a build cache's keys under build_cache/_pgp.
Within this directory is an index.json file listing all the available
keys and a <fingerprint>.pub file for each such key.

 - Adds spack.binary_distribution.generate_key_index()
   - (re)generates a build cache's key index

 - Modifies spack.binary_distribution.build_tarball()
   - if tarball is signed, automatically pushes the key used for signing
     along with the tarball
   - if regenerate_index == True, automatically (re)generates the build
     cache's key index along with the build cache's package index; as in
     spack.binary_distribution.generate_key_index()

 - Modifies spack.binary_distribution.get_keys()
   - a build cache's key index is now used instead of programmatic
     listing

 - Adds spack.binary_distribution.push_keys()
   - publishes keys from Spack's keyring to a given list of mirrors

 - Adds new spack subcommand: spack gpg publish
   - publishes keys from Spack's keyring to a given list of mirrors

 - Modifies spack.util.gpg.Gpg.signing_keys()
   - Accepts optional positional arguments for filtering the set of keys
     returned

 - Adds spack.util.gpg.Gpg.public_keys()
   - As spack.util.gpg.Gpg.signing_keys(), except public keys are
     returned

 - Modifies spack.util.gpg.Gpg.export_keys()
   - Fixes an issue where GnuPG would prompt for user input if trying to
     overwrite an existing file

 - Modifies spack.util.gpg.Gpg.untrust()
   - Fixes an issue where GnuPG would fail for input that were not key
     fingerprints

 - Modifies spack.util.web.url_exists()
   - Fixes an issue where url_exists() would throw instead of returning
     False

opadron commented:
PTAL

I've started reworking how keys are managed for build caches.  Please see the commit messages for these last two commits for details.  Please weigh in on whether you like the direction the changes are going.

Commit:
fix varmin errors

Commit:
fix tests

Commit:
fix style

Commit:
fix docs

Commit:
fix docs

Commit:
increase coverage

Commit:
add key push and fetch tests

opadron commented:
PTAL.  I've reproduced a summary of the major changes below:

> Switches to maintaining a build cache's keys under build_cache/_pgp.
Within this directory is an index.json file listing all the available
keys and a <fingerprint>.pub file for each such key.

> New spack subcommand: spack gpg publish

EDIT: ~This is ready to merge.~  Actually, I'll rebase real quick and then it'll be ready!

eugeneswalker commented:
Using this PR I was able to:
* add a key to a local filesystem build cache: `spack gpg publish -d <path> <key-id>`
* generate an key index: `spack gpg publish --rebuild-index -d <path>`
* fetch and trust the key: `spack buildcache keys -it`

opadron commented:
Update: I'm reworking some of the internal plumbing regarding how `gpg` is called by Spack.  The hope is that these changes will make a fix for the mac os test failure more straightforward.

Commit:
rework gpg module/fix error with very long GNUPGHOME dir

Commit:
style fixes

Commit:
add a shim for functools.cached_property

Commit:
handle permission denied error in gpg util

Commit:
work with older pythons

Commit:
fix tests/make gpgconf optional if no socket dir is available

Commit:
use /run/user instead of /var/run/user

Commit:
try even harder to work around gnupg issues

Commit:
fix flake8 test

Commit:
make test results even more resilient

Commit:
catch permission denied when creating /var/run/user

opadron commented:
@gartung (or anyone else who wants to review):

Could you take another pass at this?  I've had to add more changes since your last review.

gartung commented:
@eugeneswalker Can you take a look. I don't have any mirrrors up at the moment to test this.

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -293,7 +292,8 @@ def generate_package_index(cache_prefix):
         for entry in web_util.list_url(cache_prefix)
-        if entry.endswith('.yaml'))
+        if entry.endswith('.yaml')
+        and os.sep not in entry)
```
What's the extra condition about?  Is it because `list_url` is now going to iterate the mirror contents recursively?

opadron replied:
I think this extra check is not necessary, I will remove it and see if the tests still pass.

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -898,41 +948,46 @@ def get_specs():
+        keys_url = url_util.join(fetch_url,
+                                 _build_cache_relative_path,
+                                 _build_cache_keys_relative_path)
+        keys_index = url_util.join(keys_url, 'index.json')
```
Your PR description says one of the three things this PR does is update `bindist.get_keys()` so it no longer uses an index, and instead finds all files matching certain criteria under the `cache_prefix`.  But that doesn't look like what's going on here.

opadron replied:
That's my mistake.  The description documented the original intent of the PR, which has since changed.

scottwittenburg commented on lib/spack/spack/binary_distribution.py:
```diff
@@ -468,7 +509,15 @@ def build_tarball(spec, outdir, force=False, rel=False, unsigned=False,
+        # push the key to the build cache's _pgp directory so it can be
+        # imported
+        if not unsigned:
+            push_keys(outdir,
```
This is nice, if I understand correctly.  When we create a tarball and push it to a remote mirror, we will also push the key it was signed with?  I haven't got to that bit of your change yet, I'm just assuming it somehow only pushes the public part of the key.

opadron replied:
This is correct.  See `build_distribution.push_keys()` for the details.  In particular, take note of lines:

`1019` : `keys = spack.util.gpg.public_keys(*(keys or []))`

and

`1050` : `spack.util.gpg.export_keys(export_target, fingerprint)`

... the latter of which calls `gpg --export`, which is never supposed to export secret subkeys (for that, there is an entirely separate `gpg` command).

scottwittenburg commented on lib/spack/spack/test/cmd/pkg.py:
```diff
@@ -122,7 +122,8 @@ def test_pkg_add(mock_pkg_git_repo):
         try:
             assert ('A  pkg-e/package.py' in
-                    git('status', '--short', output=str))
+                    git('-c', 'color.status=off',
```
Just curious: this seems unrelated, what's going on here?

opadron replied:
Here, I was trying to work around an issue in the tests where my local git config was causing failures.

After taking a closer look and doing some googling, I realize that running git as part of a script (or in this case, as part of Spack) and *not* taking the extra measures needed to ensure that it is isolated from user- and system-provided git configs is inherently broken.  This is a separate issue that I intend to follow up on with a more-comprehensive PR.

For this PR, I will revert the change.

scottwittenburg commented on lib/spack/spack/test/stage.py:
```diff
@@ -833,7 +833,8 @@ def test_get_stage_root_in_spack(self, clear_stage_root):
                 # Make sure cached stage path value was changed appropriately
-                assert spack.stage._stage_root == test_path
+                assert spack.stage._stage_root in (
+                    test_path, os.path.join(test_path, getpass.getuser()))
```
Again, just curious, seems unrelated?

opadron replied:
This works around an issue on OSX.  I don't remember the details, so I can't be sure that there isn't a deeper issue that this change masks.  Maybe this warrants more attention?

scottwittenburg replied:
Up to you.  I didn't really intend to hold you up on this PR, but rather just try to understand what you were doing.

scottwittenburg commented on lib/spack/spack/util/gpg.py:
```diff
@@ -3,20 +3,73 @@
+    """
+    return (gnupg_home or
+            _gnupg_home_override or
+            os.getenv('SPACK_GNUPGHOME') or
```
Do you need a default of `None` here?  Like `os.getenv('SPACK_GNUPGHOME', None)`...

opadron replied:
I don't think it is necessary.

> os.getenv(varname[, value])
Return the value of the environment variable varname if it exists, or value if it doesn’t. value defaults to None.

There's no note of any version of python where they made value optional, which leads me to believe that it's always been that way.

scottwittenburg commented on lib/spack/spack/util/gpg.py:
```diff
@@ -31,43 +84,255 @@ def parse_keys_output(output):
+
+        # Make sure that the GNUPGHOME exists
+        if not os.path.exists(self.gnupg_home):
+            os.makedirs(self.gnupg_home)
```
Did you omit the `chmod 700 ...` on purpose here?

opadron replied:
Thanks for that catch.  Fixed in upcoming commits.

scottwittenburg commented on lib/spack/spack/util/gpg.py:
```diff
@@ -83,64 +348,105 @@ def create(cls, **kwargs):
+        namespace[name] = _make_wrapped_callable(func, name)
+
+
+_make_wrapped_callables(globals())
```
My $0.02: I'm not a fan of the meta-programming approach taken here, seems like a lot of complexity for the benefit of users not having to create an instance of the `Gpg` class.  It took a non-zero amount of time and effort to figure out what you were doing with the various bits: the mysterious and uncommented `wrap` decorator class that inherits from `list`, its global instance used in decorating selected methods of the `Gpg` class, the definition and subsequent deletion of this method where you generate the module level handles from stuff stored by the decorator, etc...

Again, just my opinion, but I found all that stuff kind of a pain to review, and I would have preferred something straightforward like explicit definition of the handful of module-level methods you want to expose, in terms of a singleton `Gpg` instance.  Maybe I'm just exposing myself as simple-minded here though.

opadron replied:
Not at all!  I hate it, too (and I'm the one who wrote it!).

I think there's a neat meta programming idea here that could be pulled out and stashed away in some module somewhere, but frankly I was just going for future-proofing, and sometimes I just need someone to point out when I'm being too clever for my own good!

I agree this should be simpler.  I'll take another pass and see how it turns out.

opadron replied:
Have a look at the latest commit.  Is that easier to make sense of?

scottwittenburg reviewed (approved):
I haven't actually tested this code myself, but just looked through it and provided some comments and a few questions.  Seems to make sense at a high level, and it's fixing some broken things we really need fixed, so I'm going to approve in advance of any back and forth that may happen as a result of my review comments.  Thanks @opadron!

Commit:
remove unecessary check in generate_package_index()

Commit:
revert git color workaround

Commit:
chmod 700 the user's GNUPGHOME upon creation

Commit:
simplify Gpg convenience functions
